pippy/PipelineSchedule.py

```diff
@@ -141,83 +163,108 @@ def init_p2p_neighbors(self):
         if self.is_last_stage:
             return []
-        return [dist.P2POp(dist.isend, self.fwd_output, self.next_stage)]
+        return [
+            dist.P2POp(dist.isend, fwd_output, self.next_stage)
+            for fwd_output in self.fwd_outputs
+        ]
 
-    def forward(self, microbatch: torch.Tensor):
+    def forward(self, args: List[torch.tensor]) -> torch.tensor:
         logger.info(f"[{self.rank} FORWARD {self.stage_id}")
         if self.is_first_stage:
-            self.fwd_input = microbatch
+            self.fwd_inputs = args
 
         # this is needed when we access the gradients for this in backward()
-        self.fwd_input.requires_grad = True
-        self.fwd_input.retain_grad()
+        # logger.info(f"len fwd inputs {len(self.fwd_inputs)}")
+        # logger.info(f"{self.fwd_inputs=}")
+        for tensor in self.fwd_inputs:
+            tensor.requires_grad = True
+            tensor.retain_grad()
 
         # perform forward pass on module
-        self.fwd_output = self.module(self.fwd_input)
+        self.fwd_outputs = self.module(*self.fwd_inputs)
 
         output_for_backward = (
-            self.compute_loss() if self.is_last_stage else self.fwd_output
+            self.compute_loss() if self.is_last_stage else self.fwd_outputs
         )
 
         # we store a ref to the input/output pair for this forward to be later used by the corresponding backward
         self.fwd_outputs_for_backward.append(
-            (self.fwd_input, output_for_backward)
+            (self.fwd_inputs, output_for_backward)
         )
 
-        return self.fwd_output
-
-    def get_bwd_send_ops(self) -> List[dist.P2POp]:
-        if self.is_first_stage:
-            return []
-        assert self.fwd_input.grad is not None, "grad must be valid"
-        return [dist.P2POp(dist.isend, self.fwd_input.grad, self.prev_stage)]
+        return self.fwd_outputs
 
     def get_bwd_recv_ops(self) -> List[dist.P2POp]:
         if self.is_last_stage:
             return []
-        return [dist.P2POp(dist.irecv, self.fwd_output_grads, self.next_stage)]
-
-    def sync_recv_backward_inputs(self) -> None:
-        ops = self.get_bwd_recv_ops()
-        if ops:
-            dist.batch_isend_irecv(ops).pop().wait()
+        return [
+            dist.P2POp(dist.irecv, output_grad, self.next_stage)
+            for output_grad in self.fwd_output_grads
+        ]
 
-    def _wait_backward_inputs(self):
-        assert (
-            self.bwd_recv_queue is not None
-        ), "Waiting for backward input without enqueueing one"
-        self.bwd_recv_queue.wait()
-        self.bwd_recv_queue = None
-        return self.fwd_output_grads
+    def get_bwd_send_ops(self) -> List[dist.P2POp]:
+        if self.is_first_stage:
+            return []
+        for fwd_input in self.fwd_inputs:
+            logger.info(f"{fwd_input.grad=}")
+            assert fwd_input.grad is not None, "grad must be valid"
```

Comment:
Can grad be None in some cases? e.g. some inputs do not require grad. We may want to loosen this later?

Reply:
Yes we can loosen this. Also right now even if the inputs do not require grads we send backwards a tensor of all zeros for it's gradients which is wasteful.

Reply:
I see, thanks. Do you mind putting a `# HACK` mark to describe the above presumption? Just for information purpose later.